clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -790,9 +789,108 @@ void CIRGenModule::setNonAliasAttributes(GlobalDecl GD, mlir::Operation *GO) {
   assert(!cir::MissingFeatures::setTargetAttributes());
 }
 
+static llvm::SmallVector<uint64_t> indexesOfArrayAttr(mlir::ArrayAttr indexes) {
+  llvm::SmallVector<uint64_t> inds;
+
+  for (auto i : indexes) {
+    auto ind = dyn_cast<mlir::IntegerAttr>(i);
+    assert(ind);
+    inds.push_back(ind.getValue().getSExtValue());
+  }
+
+  return inds;
+}
+
+// handle global view on the struct type
+static GlobalViewAttr createNewGlobalView(CIRGenModule &CGM, GlobalOp newGlob,
+                                          cir::StructType oldSTy,
+                                          unsigned oldInd) {
+
+  CIRGenBuilderTy &bld = CGM.getBuilder();
+  const CIRDataLayout &layout = CGM.getDataLayout();
+  mlir::MLIRContext *ctxt = bld.getContext();
+  mlir::Type newTy = newGlob.getSymType();
+  auto newPtrTy = cir::PointerType::get(ctxt, newTy);
+
+  llvm::SmallVector<int64_t> newInds;
+  uint64_t offset = oldSTy.getElementOffset(layout.layout, oldInd);
+  bld.computeGlobalViewIndicesFromFlatOffset(offset, newTy, layout, newInds);
+
+  return bld.getGlobalViewAttr(newPtrTy, newGlob, newInds);
+}
+
+// handle view on the array type
+static GlobalViewAttr
+createNewGlobalView(CIRGenModule &CGM, GlobalOp newGlob, GlobalViewAttr attr,
+                    cir::ArrayType oldTy,
+                    const llvm::SmallVector<uint64_t> &indexes) {
+
+  CIRGenBuilderTy &bld = CGM.getBuilder();
+  const CIRDataLayout &layout = CGM.getDataLayout();
+  mlir::Type newTy = newGlob.getSymType();
+  auto newArTy = dyn_cast<cir::ArrayType>(newTy);
+  assert(newArTy);
```

Comment:
Looks like `newArTy` is only used for the assert? If so better to do the dyn_cast directly inside the assert